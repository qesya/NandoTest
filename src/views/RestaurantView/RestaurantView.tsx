import React, {useCallback} from 'react';
import {
  StyleSheet,
  View,
  Text,
  FlatList,
  Linking,
  StatusBar,
} from 'react-native';
import * as colours from '../../utilities/colours';
import * as fonts from '../../utilities/fonts';
import RestaurantCard from '../../components/RestaurantCard/RestaurantCard';
import {Restaurant} from '../../utilities/response';
import {SafeAreaView} from 'react-native-safe-area-context';
import Button from '../../components/Button/Button';

type Props = {
  data: Array<Restaurant>;
  loading: boolean;
  onPressGoBack: () => void;
};

interface RenderItem<T> {
  item: T;
}

const RestaurantView = (props: Props) => {
  const openUrlHandler = useCallback(
    (url: string) =>
      Linking.openURL(url).catch(err =>
        console.error('An error occurred when opening url', err),
      ),
    [],
  );

  const renderRestaurantItem = useCallback(
    ({item}: RenderItem<Restaurant>) => {
      return (
        <>
          <RestaurantCard
            onPressRestaurant={() => openUrlHandler(item.url ?? '')}
            restaurantName={item.name}
            addressLocality={item?.geo?.address?.addressLocality}
            address={item?.geo?.address?.streetAddress}
            postalCode={item?.geo?.address?.postalCode}
          />
        </>
      );
    },
    [openUrlHandler],
  );
  return (
    <SafeAreaView style={styles.container}>
      <StatusBar
        translucent
        backgroundColor={colours.WHITE}
        barStyle="dark-content"
      />
      <View style={styles.innerContainer}>
        <Text style={styles.txtHeader}>Nando's Restaurants</Text>
        {!props.loading ? (
          <FlatList
            testID="flatlist-restaurant"
            automaticallyAdjustsScrollIndicatorInsets
            style={[styles.restaurantList]}
            showsVerticalScrollIndicator={false}
            contentContainerStyle={styles.containerFlatlist}
            data={props.data}
            keyExtractor={item => item.name}
            renderItem={renderRestaurantItem}
            contentInsetAdjustmentBehavior="automatic"
            ItemSeparatorComponent={() => <View style={styles.itemSeparator} />}
          />
        ) : (
          <Text style={styles.txtHeader}>Loading...</Text>
        )}
        <Button style={styles.btn} title="Back" onPress={props.onPressGoBack} />
      </View>
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: colours.WHITE,
  },
  innerContainer: {
    flex: 1,
    paddingHorizontal: 16,
  },
  txtHeader: {
    ...fonts.PoppinsLight,
    fontSize: 24,
    textAlign: 'center',
    paddingHorizontal: 20,
    marginVertical: 20,
    color: colours.BLACK,
    textTransform: 'uppercase',
    letterSpacing: 2,
  },
  restaurantList: {
    flex: 1,
  },
  containerFlatlist: {
    paddingBottom: 30,
    paddingHorizontal: -40,
  },
  itemSeparator: {
    height: 20,
  },
  btn: {
    marginVertical: 20,
  },
});
export default RestaurantView;
