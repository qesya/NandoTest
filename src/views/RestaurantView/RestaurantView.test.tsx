import React from 'react';
import renderer from 'react-test-renderer';
import RestaurantView from './RestaurantView';
import {render} from '@testing-library/react-native';

describe('RestaurantView', () => {
  test('renders correctly', () => {
    const component = renderer
      .create(
        <RestaurantView
          onPressGoBack={() => jest.fn()}
          data={[]}
          loading={false}
        />,
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });
  test('render loading', () => {
    const component = render(
      <RestaurantView
        onPressGoBack={() => jest.fn()}
        data={[]}
        loading={true}
      />,
    );
    expect(component.queryByTestId('LoadingText')).toBeDefined();
  });
});
