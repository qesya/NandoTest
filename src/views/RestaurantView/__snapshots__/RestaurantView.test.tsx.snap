// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RestaurantView renders correctly 1`] = `
<RNCSafeAreaView
  edges={
    Array [
      "bottom",
      "left",
      "right",
      "top",
    ]
  }
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "paddingHorizontal": 16,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#1F1F1F",
          "fontFamily": "Poppins-Light",
          "fontSize": 24,
          "letterSpacing": 2,
          "marginVertical": 20,
          "paddingHorizontal": 20,
          "textAlign": "center",
          "textTransform": "uppercase",
        }
      }
    >
      Nando's Restaurants
    </Text>
    <RCTScrollView
      ItemSeparatorComponent={[Function]}
      automaticallyAdjustsScrollIndicatorInsets={true}
      contentContainerStyle={
        Object {
          "paddingBottom": 30,
          "paddingHorizontal": -40,
        }
      }
      contentInsetAdjustmentBehavior="automatic"
      data={Array []}
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      showsVerticalScrollIndicator={false}
      stickyHeaderIndices={Array []}
      style={
        Array [
          Object {
            "flex": 1,
          },
        ]
      }
      testID="flatlist-restaurant"
      viewabilityConfigCallbackPairs={Array []}
    >
      <View />
    </RCTScrollView>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "alignItems": "center",
            "alignSelf": "stretch",
            "backgroundColor": "#1F1F1F",
            "borderRadius": 40,
            "height": 40,
            "justifyContent": "center",
            "marginBottom": 10,
          },
          Object {
            "marginVertical": 20,
          },
        ]
      }
    >
      <Text
        style={
          Object {
            "color": "#FFFFFF",
            "fontFamily": "Poppins-Light",
            "fontSize": 14,
          }
        }
      >
        Back
      </Text>
    </View>
  </View>
</RNCSafeAreaView>
`;
